File: frontend/src/utils/memoryPipeline.ts
export const SEGMENT_SIZE = 10
export const PER_SEGMENT_MERGE_MAX_ROUND = 40
export const BATCH_MERGE_SIZE = 4

export interface SegmentRange {
  start: number
  end: number
}

export interface HistorySegmentItem {
  id: number
  start_round: number
  end_round: number
  history_segment: string
}

export type PipelineMergeMode = 'single' | 'batch' | 'forced_tail'

export interface PipelineCompressStep {
  type: 'compress'
  segment: SegmentRange
  segmentIndex: number
}

export interface PipelineMergeStep {
  type: 'merge'
  segments: SegmentRange[]
  mergeMode: PipelineMergeMode
}

export type PipelineStep = PipelineCompressStep | PipelineMergeStep

export interface PipelinePlan {
  rangeStart: number
  rangeEnd: number
  segments: SegmentRange[]
  steps: PipelineStep[]
  compressCount: number
  mergeCount: number
  hasForcedTailMerge: boolean
}

/** 按全局 10 轮网格切分，与用户选定范围取交集 */
export function splitGlobalSegments(rangeStart: number, rangeEnd: number): SegmentRange[] {
  const start = Math.floor(rangeStart)
  const end = Math.floor(rangeEnd)
  if (start > end) return []

  const segments: SegmentRange[] = []
  let gridStart = Math.floor((start - 1) / SEGMENT_SIZE) * SEGMENT_SIZE + 1

  while (gridStart <= end) {
    const gridEnd = gridStart + SEGMENT_SIZE - 1
    const segStart = Math.max(gridStart, start)
    const segEnd = Math.min(gridEnd, end)
    if (segStart <= segEnd) {
      segments.push({ start: segStart, end: segEnd })
    }
    gridStart += SEGMENT_SIZE
  }

  return segments
}

function segmentRangeLabel(segments: SegmentRange[]): string {
  if (segments.length === 0) return ''
  if (segments.length === 1) {
    const seg = segments[0]
    return `${seg.start}-${seg.end}轮`
  }
  const first = segments[0]
  const last = segments[segments.length - 1]
  return `${first.start}-${last.end}轮（${segments.length}段）`
}

function pushMergeStep(
  steps: PipelineStep[],
  segments: SegmentRange[],
  mergeMode: PipelineMergeMode,
): void {
  if (segments.length === 0) return
  steps.push({
    type: 'merge',
    segments: segments.map((seg) => ({ ...seg })),
    mergeMode,
  })
}

/**
 * 模拟运行时合并策略，生成 compress / merge 步骤序列。
 * 40 轮后攒满 4 段批量合并；末尾不足 4 段时追加 forced_tail 合并。
 */
export function planMemoryPipeline(rangeStart: number, rangeEnd: number): PipelinePlan {
  const segments = splitGlobalSegments(rangeStart, rangeEnd)
  const steps: PipelineStep[] = []
  let pendingBatch: SegmentRange[] = []
  let hasForcedTailMerge = false

  segments.forEach((segment, index) => {
    steps.push({
      type: 'compress',
      segment: { ...segment },
      segmentIndex: index + 1,
    })

    if (segment.end <= PER_SEGMENT_MERGE_MAX_ROUND) {
      pushMergeStep(steps, [segment], 'single')
      return
    }

    pendingBatch.push({ ...segment })
    if (pendingBatch.length >= BATCH_MERGE_SIZE) {
      pushMergeStep(steps, pendingBatch, 'batch')
      pendingBatch = []
    }
  })

  if (pendingBatch.length > 0) {
    hasForcedTailMerge = pendingBatch.length < BATCH_MERGE_SIZE
    pushMergeStep(steps, pendingBatch, hasForcedTailMerge ? 'forced_tail' : 'batch')
    pendingBatch = []
  }

  return {
    rangeStart,
    rangeEnd,
    segments,
    steps,
    compressCount: segments.length,
    mergeCount: steps.filter((step) => step.type === 'merge').length,
    hasForcedTailMerge,
  }
}

export function pipelinePlanSummary(plan: PipelinePlan): string {
  const parts = [`${plan.compressCount} 段压缩`, `${plan.mergeCount} 次合并`]
  if (plan.hasForcedTailMerge) {
    parts.push('含 1 次强制尾批')
  }
  return parts.join(' + ')
}

export function pipelineStepLabel(step: PipelineStep): string {
  if (step.type === 'compress') {
    const { start, end } = step.segment
    return `Segment 压缩 · 第 ${start}-${end} 轮（段 ${step.segmentIndex}）`
  }

  const label = segmentRangeLabel(step.segments)
  if (step.mergeMode === 'single') {
    return `History 合并 · 单段 · ${label}`
  }
  if (step.mergeMode === 'batch') {
    return `History 合并 · 批量 ${step.segments.length} 段 · ${label}`
  }
  return `History 合并 · 强制尾批 ${step.segments.length} 段 · ${label}`
}

function roundRangeText(start: number, end: number): string {
  return start === end ? `第 ${start} 轮` : `第 ${start}-${end} 轮`
}

/** 将相邻的 compress + merge 合并为一句，用于管线预览 */
export function formatPipelineCycleLines(plan: PipelinePlan): string[] {
  const lines: string[] = []
  let index = 0

  while (index < plan.steps.length) {
    const step = plan.steps[index]
    if (step.type !== 'compress') {
      index += 1
      continue
    }

    const next = plan.steps[index + 1]
    const { start, end } = step.segment
    const roundLabel = roundRangeText(start, end)

    if (
      next?.type === 'merge' &&
      next.mergeMode === 'single' &&
      next.segments.length === 1 &&
      next.segments[0].start === start &&
      next.segments[0].end === end
    ) {
      lines.push(`${roundLabel}：Segment 压缩后单段合并`)
      index += 2
      continue
    }

    const compressSteps: PipelineCompressStep[] = [step]
    let cursor = index + 1
    while (cursor < plan.steps.length && plan.steps[cursor].type === 'compress') {
      compressSteps.push(plan.steps[cursor] as PipelineCompressStep)
      cursor += 1
    }

    const mergeStep = plan.steps[cursor]
    if (mergeStep?.type === 'merge') {
      const first = compressSteps[0].segment
      const last = compressSteps[compressSteps.length - 1].segment
      const batchRoundLabel = roundRangeText(first.start, last.end)
      if (mergeStep.mergeMode === 'forced_tail') {
        lines.push(
          `${batchRoundLabel}：${compressSteps.length} 段 Segment 压缩后强制合并（尾批不足 4 段）`,
        )
      } else {
        lines.push(`${batchRoundLabel}：${compressSteps.length} 段 Segment 压缩后批量合并`)
      }
      index = cursor + 1
      continue
    }

    lines.push(`${roundLabel}：Segment 压缩`)
    index += 1
  }

  return lines
}

File: server/memory_pipeline.py
"""RP 链路测试分段与合并计划（与 frontend memoryPipeline.ts 对齐）。"""
from __future__ import annotations

from copy import deepcopy
from typing import Any, Literal, TypedDict, Union

FIRST_SEGMENT_ROUNDS = 10
BATCH_MERGE_SIZE = 4

PipelineMergeMode = Literal["single", "batch", "forced_tail"]


class SegmentRange(TypedDict):
    start: int
    end: int


class PipelineCompressStep(TypedDict):
    type: Literal["compress"]
    segment: SegmentRange
    segmentIndex: int


class PipelineMergeStep(TypedDict):
    type: Literal["merge"]
    segments: list[SegmentRange]
    mergeMode: PipelineMergeMode


PipelineStep = Union[PipelineCompressStep, PipelineMergeStep]


class PipelinePlan(TypedDict):
    rangeStart: int
    rangeEnd: int
    segments: list[SegmentRange]
    steps: list[PipelineStep]
    compressCount: int
    mergeCount: int
    hasForcedTailMerge: bool


def segment_range_for_index(segment_index: int) -> SegmentRange:
    if segment_index < 1:
        raise ValueError("segment_index must be >= 1")
    return {
        "start": (segment_index - 1) * FIRST_SEGMENT_ROUNDS + 1,
        "end": segment_index * FIRST_SEGMENT_ROUNDS,
    }


def split_global_segments(range_start: int, range_end: int) -> list[SegmentRange]:
    start = int(range_start)
    end = int(range_end)
    if start > end:
        return []

    segments: list[SegmentRange] = []
    index = 1
    while True:
        full = segment_range_for_index(index)
        if full["start"] > end:
            break
        if full["end"] >= start:
            seg_start = full["start"]
            seg_end = min(full["end"], end)
            if seg_start <= seg_end:
                segments.append({"start": seg_start, "end": seg_end})
        index += 1
    return segments


def _push_merge_step(
    steps: list[PipelineStep],
    segments: list[SegmentRange],
    merge_mode: PipelineMergeMode,
) -> None:
    if not segments:
        return
    steps.append(
        {
            "type": "merge",
            "segments": [deepcopy(seg) for seg in segments],
            "mergeMode": merge_mode,
        }
    )


def plan_memory_pipeline(range_start: int, range_end: int) -> PipelinePlan:
    segments = split_global_segments(range_start, range_end)
    steps: list[PipelineStep] = []
    pending_batch: list[SegmentRange] = []
    has_forced_tail_merge = False

    for index, segment in enumerate(segments):
        steps.append(
            {
                "type": "compress",
                "segment": deepcopy(segment),
                "segmentIndex": index + 1,
            }
        )
        pending_batch.append(deepcopy(segment))
        if len(pending_batch) >= BATCH_MERGE_SIZE:
            _push_merge_step(steps, pending_batch, "batch")
            pending_batch = []

    if pending_batch:
        has_forced_tail_merge = len(pending_batch) < BATCH_MERGE_SIZE
        _push_merge_step(
            steps,
            pending_batch,
            "forced_tail" if has_forced_tail_merge else "batch",
        )

    return {
        "rangeStart": range_start,
        "rangeEnd": range_end,
        "segments": segments,
        "steps": steps,
        "compressCount": len(segments),
        "mergeCount": sum(1 for step in steps if step["type"] == "merge"),
        "hasForcedTailMerge": has_forced_tail_merge,
    }


def pipeline_step_label(step: PipelineStep) -> str:
    if step["type"] == "compress":
        seg = step["segment"]
        return f"Segment 压缩 · 第 {seg['start']}-{seg['end']} 轮（段 {step['segmentIndex']}）"

    segments = step["segments"]
    if len(segments) == 1:
        label = f"{segments[0]['start']}-{segments[0]['end']}轮"
    else:
        label = f"{segments[0]['start']}-{segments[-1]['end']}轮（{len(segments)}段）"

    mode = step["mergeMode"]
    if mode == "single":
        return f"History 合并 · 单段 · {label}"
    if mode == "batch":
        return f"History 合并 · 批量 {len(segments)} 段 · {label}"
    return f"History 合并 · 强制尾批 {len(segments)} 段 · {label}"
